src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -562,19 +590,21 @@ private void RecordWaiterStartTime()
                 // Don't record zero, that value is reserved for indicating that a time is not recorded
                 currentTimeMs--;
             }
-            _waiterStartTimeMs = currentTimeMs;
+            WaiterStartTimeMs = currentTimeMs;
         }
 
         private bool ShouldStopPreemptingWaiters
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
             get
             {
-                // If the recorded time is zero, a time has not been recorded yet
-                ushort waiterStartTimeMs = _waiterStartTimeMs;
+                // If the recorded time is zero, a time has not been recorded yet. The stored waiter start time (ms as a ushort)
+                // excludes the upper bit since the field is also used for a flag, so also mask off the upper bits from the
+                // comparison with the current tick count.
+                ushort waiterStartTimeMs = WaiterStartTimeMs;
                 return
                     waiterStartTimeMs != 0 &&
-                    (ushort)Environment.TickCount - waiterStartTimeMs >= MaxDurationMsForPreemptingWaiters;
+                    (ushort)((Environment.TickCount - waiterStartTimeMs) & 0x7fff) >= MaxDurationMsForPreemptingWaiters;
```

Comment:
```suggestion
                    ((Environment.TickCount - waiterStartTimeMs) & 0x7fff) >= MaxDurationMsForPreemptingWaiters;
```
The cast to ushort is unnecessary.

I think that the `ushort WaiterStartTimeMs` property is misleading given that only 15 bits are actually valid. The type of the property can be `int` with a comment that only lower 15 bits are valid. Nearly all callers need to be aware of it.

There is another subtle bug a few lines above:
```
            if (currentTimeMs == 0)
            {
                // Don't record zero, that value is reserved for indicating that a time is not recorded
                currentTimeMs--;
            }
```
This won't work correctly when currentTimeMs is `0x8000`. It is going to be recorded as 0 that is not correct.

Reply:
Thanks again, fixed

Reply:
I wonder if using the lowermost bit as the flag (as in https://github.com/dotnet/runtime/pull/97227#discussion_r1463723951) would avoid most of the trouble and make this simpler?

At least `ushort WaiterStartTimeMs` would not be misleading as the whole ushort range would be in use. 
Also wrap around would happen in 60 sec. not in 30.

Reply:
```cs
        private ushort WaiterStartTimeMs
        {
            get => (ushort)(_waiterStartTimeMsAndFlags &  ~1);
            set => _waiterStartTimeMsAndFlags = (ushort)(value | (_waiterStartTimeMsAndFlags & 1));
        }
```

Then the same comparison as before might work:

```cs
(ushort)Environment.TickCount - WaiterStartTimeMs >= MaxDurationMsForPreemptingWaiters;
```

Reply:
That would work, but would need to make sure that 0 is not recorded when the actual time is 1, so there's still a little bit of complication at the caller side. I was thinking I could just use a bit in the `_state` field to simplify. The waiter count has plenty of bits and one could be used for this instead. The waiter start time is stored and used in only a couple of places so it's probably ok as is, but I'll go ahead and make that change to use a bit in the `_state` field, it would be a bit cleaner.

Reply:
Also the previous code when the ushort field was being used directly:
> `(ushort)Environment.TickCount - waiterStartTimeMs >= MaxDurationMsForPreemptingWaiters`
Doesn't work because of the implicit int promotion in the subtract. Eg. if the recorded start time was `(ushort)-100` and the current time as a ushort is 0, the result of the subtract would be a large negative int value instead of 100, and would not trigger the heuristic. Need additional casting here to force ushort math.

Reply:
Maybe using `short` instead would simplify, as the int promotion would sign-extend instead, will consider

Reply:
That wouldn't work either, this expression would need extra casting it seems